crates/vm/src/arch/vm.rs

```diff
@@ -154,8 +170,8 @@ where
         &self,
         exe: impl Into<VmExe<F>>,
         input: impl Into<Streams<F>>,
-        mut f: impl FnMut(usize, ExecutionSegment<F, VC>) -> R,
-    ) -> Result<Vec<R>, ExecutionError> {
+        mut f: impl FnMut(usize, ExecutionSegment<F, VC>) -> Result<R, GenerationError>,
```

Comment:
nit: could we make this function generic in `R, E` and have `f: impl FnMut(..) -> Result<R, E>` and the overall function `-> Result<Vec<R>, E>` and take another closure `map_err: impl Fn(ExecutionError) -> E`?

It seems a little better and means we can localize `GenerationPipelineError` (or even better remove it) to only the place where we use it

Reply:
with this change, hoping we can just make `GenerationError` contain `ExecutionError` as a `#[from]`